### ISSUE TITLE
Issue with Service Discovery in K8s with "NotReady" node

### ISSUE THREAD

**Issue by uromahn** [jgroups-extras/jgroups-kubernetes#49]
We are using Infinispan in our application which gets deployed to a Kubernetes (k8s) cluster. In our application, we are using KUBE_PING for service discovery.

There is an edge case when the Infinispan cache cluster won't properly synchronize causing all new pods that get started or existing to get restarted to fail and go into a restart-loop.

Here is the situation:
In case a pod is running on a compute node that is being taken offline and now going into status "NotReady", the pod on that node will become unavailable and its status will be changed to "Unknown" making it unavailable for incoming requests. K8s will then schedule and start a new pod on a different node which will eventually be in status "Ready".

However, when querying the k8s API for all pods in a namespace, the response will now contain all running pods plus the disabled pod with status "Unknown".

When a new pod with an infinispan cluster gets started or an existing gets re-started, the KUBE_PING discovery from JGroups will get all pod IPs via the k8s API, regardless of the pod status!

Infinispan now tries to reach all pods in that list to establish a connection and join the cluster. However, the pod on the "NotReady" node will never respond and as a result, our Infinispan cluster will never be able to synchronize and consequently the starting pods will go into an infinite restart loop.

I plan to make a change to the existing KUBE_PING and add the status of all container running in a pod to the Pod.java object. Only pods where all container have a status of "ready" will get added to the list in KUBE_PING.

**Comment by belaban** [comment]
Hmm... is this similar to the old version of KUBE_PING where we queried only pods that were in `Ready` state [1]?

I recall that we changed this once to only query the `Running` pods, how is this different from `Ready`?

[1] https://github.com/jgroups-extras/jgroups-kubernetes/blob/stable/kube/src/main/java/org/jgroups/ping/kube/Client.java#L118

**Comment by uromahn** [comment]
I am not in front of my computer now. I will check the code tomorrow my
time and update the ticket.

**Comment by slaskawi** [comment]
@belaban I believe it depends on the version. I recall removing the "ready" check from the latest version (https://github.com/jgroups-extras/jgroups-kubernetes/blame/master/src/main/java/org/jgroups/protocols/kubernetes/Client.java#L98-L118) since I couldn't find a use case where Kubernetes assigns an IP Address to a Pod but keeps it in a now `Running` state.

Therefore, I would really like to learn when does it happen. @uromahn May I kindly ask you to send a full Kubernetes response obtained from https://github.com/jgroups-extras/jgroups-kubernetes/blame/master/src/main/java/org/jgroups/protocols/kubernetes/Client.java#L66 ?

**Comment by belaban** [comment]
@slaskawi Yes, I recall we had a discussion on that, but let's see what @uromahn has to say...

**Comment by uromahn** [comment]
I am trying to reproduce the situation which is not easy with a K8s cluster consisting of almost 100 nodes. I may have to setup a local test environment with just two nodes hoping I can reproduce there.

**Comment by uromahn** [comment]
In the meantime, I did some research...
We are using Infinispan 9.1.4 (soon to be upgraded to 9.2.5) which is using jgroups-kubernetes 1.0.5.Final. We are running it on a K8s 1.6.13 cluster with currently 69 node (however, in production we have 300+ nodes!).

In the `Client.java`, I found the following comment in the method `parseJsonResult()`:
```Java
                //Previously we did checks on phase. But from my observations, it is extremely rare to have a container
                //listed by Kubernetes API with any other status but Running (I might imagine it will hang in scheduled).
                //However in both cases, its IP address will be null. So it is much better to stick to that.
```

This is causing us grieve.
The observed issue happens because of this reported K8s issue: https://github.com/kubernetes/kubernetes/issues/51333

So, when a node where one of our pods was running suddenly "disappears", the K8s API server does not remove the pod from its internal database and just sets the status to "Unknown". Once the lifeness probe doesn't respond, K8s will schedule a new pod on a different node and the "Unknown" pod will be taken out of rotation for request routing. However, all the previous metadata of the "Unknown" pod will still be in etcd and will get returned from a call to the API server. The returned metadata also contains the podIp of the node that just went down.
So, when a new pod containing an Infinispan node comes up, it will still "see" the "Unknown" pod and its IP address and will add it to its internal list of participating nodes in KUBE_PING. However, based on our current setup, the Infinispan cache synchronization in the new pod will never succeed since any attempt to reach the "dead" node will time out - there is nobody listening on that IP anymore. Unfortunately, our application expects to write something to the cache synchronously during startup and if that times out - which it does - it never has its readiness probe up and after some time, K8s will kill the pod and restart it with the same outcome. So, we end up in an infinite restart loop.

While our issue is partially "home made" due to the way we start our application relying on successful cache synchronization, the issue nevertheless exists and the client in KUBE_PING should never return "unhealthy" pods, even if in 99% of the cases those pods won't have an IP address. But, as you can see, there are some edge cases when the IP address of unhealthy pods will get retained in etcd and we should not report them back to KUBE_PING in the list of pods.

**Comment by uromahn** [comment]
Sorry for the apparent late response. I am in California, 9 hours behind you guys. My GitHub location wasn't correctly updated.

**Comment by slaskawi** [comment]
@uromahn @belaban Ok, so I guess the solution is clear - we need to check if the pod is in `Running` status.

**Comment by uromahn** [comment]
@belaban and @slaskawi - I have started working on such a solution. I have also quite a few example JSON files which I am going to "sanitize" and use for unit tests that simulates some different scenarios.
Unfortunately, just checking for `Running` status isn't going to be enough. I'll explain more in the code as comments once I have finished it.
I am going on a two week vacation starting Saturday and likely won't have time to complete my change. I will complete this the week starting July 9th and then create a pull request for you to review (and hopefully approve).

**Comment by belaban** [comment]
OK, sounds good. Note that JGroups can do discovery in parallel, and thus won't be blocked by an unresponsive member, plus such a member would eventually be excluded from the cluster view if it doesn't send out heartbeats. But that's probably too late in the process for you

**Comment by uromahn** [comment]
Closed due to fix merged into master.
